forester/src/nullifier/address/processor.rs

```diff
@@ -0,0 +1,324 @@
+                            for next_stage in next_stages {
+                                self.output.send(next_stage).await.unwrap();
+                            }
+
+                        }
+                        Err(e) => {
+                            warn!("Error in AddressProcessor: {:?}", e);
+                        }
+                    }
+                }
+                _ = self.close_output.recv() => {
+                    debug!("Received signal to close output channel");
+                    break;
+                }
+                else => break,
+            }
+            if self.shutdown.load(Ordering::Relaxed) {
+                debug!("Shutdown signal received, stopping AddressProcessor");
+                break;
+            }
+        }
+        debug!("AddressProcessor process completed");
+    }
+
+    async fn fetch_address_queue_data(
+        &self,
+        context: &PipelineContext<T, R>,
+    ) -> Result<Vec<AddressPipelineStage<T, R>>, ForesterError> {
+        let config = &context.config;
+        let rpc = &context.rpc;
+
+        let queue_data = fetch_address_queue_data(config, rpc).await?;
+        if queue_data.is_empty() {
+            debug!("Address queue is empty");
+            Ok(vec![AddressPipelineStage::Complete])
+        } else {
+            Ok(vec![AddressPipelineStage::ProcessAddressQueue(
+                context.clone(),
+                queue_data,
+            )])
+        }
+    }
+
+    async fn process_address_queue(
+        &self,
+        context: PipelineContext<T, R>,
+        queue_data: Vec<ForesterQueueAccount>,
+    ) -> Result<Vec<AddressPipelineStage<T, R>>, ForesterError> {
+        let mut next_stages = Vec::new();
+        for account in queue_data {
+            next_stages.push(AddressPipelineStage::UpdateAddressMerkleTree(
+                context.clone(),
+                account,
+            ));
+        }
+        Ok(next_stages)
+    }
+
+    async fn update_address_merkle_tree(
+        &self,
+        context: PipelineContext<T, R>,
+        account: ForesterQueueAccount,
+    ) -> Result<Vec<AddressPipelineStage<T, R>>, ForesterError> {
+        let config = &context.config;
+        let rpc = &context.rpc;
+        let indexer = &context.indexer;
+
+        let address = account.hash;
+        let address_hashset_index = account.index;
+
+        let proof = indexer
+            .lock()
+            .await
+            .get_multiple_new_address_proofs(config.address_merkle_tree_pubkey.to_bytes(), address)
+            .await
+            .map_err(|_| ForesterError::Custom("Failed to get address tree proof".to_string()))?;
+        // TODO: use changelog array size from tree config
+        let changelog = proof.root_seq % ADDRESS_MERKLE_TREE_CHANGELOG;
+        // TODO: add index changelog current changelog index to the proof or we make them the same size
+        let index_changelog = proof.root_seq % ADDRESS_MERKLE_TREE_CHANGELOG;
```

Comment:
this should be `ADDRESS_MERKLE_TREE_INDEXED_CHANGELOG` default value is 256 right now.